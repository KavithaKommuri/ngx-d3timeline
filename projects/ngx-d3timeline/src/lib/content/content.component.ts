import { Component } from '@angular/core';
import { ContentService } from './content.service';

@Component({
  selector: '[ngx-d3timeline-content]',
  template: `
    <svg:g
      class="content-group"
      *ngIf="contentService.eventRectangles$ | async as eventRectangles"
    >
      <svg:g
        *ngFor="let eventRectangle of eventRectangles"
        [attr.transform]="eventRectangle.transform"
      >
        <svg:rect
          [attr.height]="eventRectangle.height"
          [attr.width]="eventRectangle.width"
        ></svg:rect>
        <svg:text dy="1em">{{ eventRectangle.title }}</svg:text>
      </svg:g>
    </svg:g>
  `,
  styles: [
    `
      .content-group {
        font-size: 10px;
      }

      rect {
        fill: none;
        stroke: #000;
      }
    `
  ]
})
export class ContentComponent {
  constructor(public contentService: ContentService) {}
}
