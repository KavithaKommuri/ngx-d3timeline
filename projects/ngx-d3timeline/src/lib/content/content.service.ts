import { Injectable } from '@angular/core';
import { ScalesService } from '../scales.service';
import { State } from '../state';
import { ScaleBand, ScaleTime } from 'd3-scale';
import { map } from 'rxjs/operators';
import { TimelineEvent } from '../timeline-event';
import { EventRectangle } from './content';
import { Orientation } from '../orientation';

@Injectable({ providedIn: 'root' })
export class ContentService {
  eventRectangles$ = this.scalesService.scales$.pipe(
    map(scales => this.createEventRectangles(scales))
  );

  constructor(private scalesService: ScalesService) {}

  createEventRectangles({
    scaleBand,
    scaleTime,
    state
  }: {
    scaleBand: ScaleBand<string>;
    scaleTime: ScaleTime<number, number>;
    state: State;
  }): EventRectangle[] {
    return state.data.map(d => ({
      title: d.type,
      transform: this.dataTransform(
        d,
        state.axisOrientations.timeOrientation,
        scaleBand,
        scaleTime
      ),
      width: this.rectWidth(
        d,
        state.axisOrientations.timeOrientation,
        scaleBand,
        scaleTime
      ),
      height: this.rectHeight(
        d,
        state.axisOrientations.timeOrientation,
        scaleBand,
        scaleTime
      )
    }));
  }

  dataTransform(
    data: TimelineEvent,
    orientation: Orientation,
    scaleBand: ScaleBand<string>,
    scaleTime: ScaleTime<number, number>
  ) {
    return `translate(${this.getEventX(
      data,
      orientation,
      scaleBand,
      scaleTime
    )}, ${this.getEventY(data, orientation, scaleBand, scaleTime)})`;
  }

  rectHeight(
    data: TimelineEvent,
    orientation: Orientation,
    scaleBand: ScaleBand<string>,
    scaleTime: ScaleTime<number, number>
  ) {
    return orientation === Orientation.Vertical
      ? this.rectTimeBreadth(data, scaleTime)
      : this.rectResourceBreadth(scaleBand);
  }

  rectWidth(
    data: TimelineEvent,
    orientation: Orientation,
    scaleBand: ScaleBand<string>,
    scaleTime: ScaleTime<number, number>
  ) {
    return orientation === Orientation.Vertical
      ? this.rectResourceBreadth(scaleBand)
      : this.rectTimeBreadth(data, scaleTime);
  }
  private getEventX(
    data: TimelineEvent,
    orientation: Orientation,
    scaleBand: ScaleBand<string>,
    scaleTime: ScaleTime<number, number>
  ) {
    return orientation === Orientation.Vertical
      ? this.positionInResourceAxis(data, scaleBand)
      : this.positionInTimeAxis(data, scaleTime);
  }

  private getEventY(
    data: TimelineEvent,
    orientation: Orientation,
    scaleBand: ScaleBand<string>,
    scaleTime: ScaleTime<number, number>
  ) {
    return orientation === Orientation.Vertical
      ? this.positionInTimeAxis(data, scaleTime)
      : this.positionInResourceAxis(data, scaleBand);
  }

  private positionInResourceAxis(
    data: TimelineEvent,
    scaleBand: ScaleBand<string>
  ): number {
    return scaleBand(data.series);
  }

  private positionInTimeAxis(
    data: TimelineEvent,
    scaleTime: ScaleTime<number, number>
  ): number {
    return scaleTime(data.start);
  }

  private rectTimeBreadth(
    data: TimelineEvent,
    scaleTime: ScaleTime<number, number>
  ): number {
    return scaleTime(data.finish) - scaleTime(data.start);
  }

  private rectResourceBreadth(scaleBand: ScaleBand<string>): number {
    return scaleBand.bandwidth();
  }
}
