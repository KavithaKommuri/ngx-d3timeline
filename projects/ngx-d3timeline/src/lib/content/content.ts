export interface EventRectangle {
  title: string;
  transform: string;
  width: number;
  height: number;
}
