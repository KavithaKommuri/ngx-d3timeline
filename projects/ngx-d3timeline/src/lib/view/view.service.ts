import { Injectable } from '@angular/core';

@Injectable({ providedIn: 'root' })
export class ViewService {
  readonly margin = 50;

  get rootTransform() {
    return `translate(${this.margin}, ${this.margin})`;
  }
}
