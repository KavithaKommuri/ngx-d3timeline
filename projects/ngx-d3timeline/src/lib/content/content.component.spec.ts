import { TestBed, ComponentFixture } from '@angular/core/testing';
import { TimelineView } from '../view/timeline-view';
import { mockData } from '../mock-data';
import { Orientation } from '../orientation';
import { ContentComponent } from './content.component';
import { ResourcesAxisViewModel } from '../axis/resources-axis/resources-axis-view-model';
import { TimeAxisViewModel } from '../axis/time-axis/time-axis-view-model';
import { ContentService } from './content.service';
import { of } from 'rxjs';
import { EventRectangle } from './content';

describe('ContentComponent', () => {
  let fixture: ComponentFixture<ContentComponent>;
  let contentService: ContentService;

  beforeEach(() => {
    TestBed.configureTestingModule({
      declarations: [ContentComponent],
      providers: [{ provide: ContentService, useValue: jest.fn() }]
    });

    fixture = TestBed.createComponent(ContentComponent);
    contentService = TestBed.inject(ContentService);
  });

  it('should not render if eventRectangles are null', () => {
    contentService.eventRectangles$ = of(null);
    fixture.detectChanges();

    expect(fixture.nativeElement).toMatchSnapshot();
  });

  it('should render correctly', () => {
    const rectangles: EventRectangle[] = [
      {
        title: 'Event 1',
        transform: 'translate(50,0)',
        width: 50,
        height: 80
      },
      {
        title: 'Event 2',
        transform: 'translate(130,0)',
        width: 60,
        height: 90
      }
    ];
    contentService.eventRectangles$ = of(rectangles);
    fixture.detectChanges();

    expect(fixture.nativeElement).toMatchSnapshot();
  });
});
