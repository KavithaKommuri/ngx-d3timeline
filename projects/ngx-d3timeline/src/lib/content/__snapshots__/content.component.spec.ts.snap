// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContentComponent should not render if eventRectangles are null 1`] = `
<div
  id="root0"
  ngx-d3timeline-content=""
>
  
</div>
`;

exports[`ContentComponent should render correctly 1`] = `
<div
  id="root1"
  ngx-d3timeline-content=""
>
  
  <g
    class="content-group"
  >
    
    <g
      transform="translate(50,0)"
    >
      <rect
        height="80"
        width="50"
      />
      <text
        dy="1em"
      >
        Event 1
      </text>
    </g>
    <g
      transform="translate(130,0)"
    >
      <rect
        height="90"
        width="60"
      />
      <text
        dy="1em"
      >
        Event 2
      </text>
    </g>
  </g>
</div>
`;
